shared/src/main/scala/mlscript/TypeSimplifier.scala

```diff
@@ -282,4 +282,107 @@ trait TypeSimplifier { self: Typer =>
     
   }
   
+  def reconstructClassTypes(st: SimpleType, pol: Bool, ctx: Ctx): SimpleType = {
+    
+    val renewed = MutMap.empty[TypeVariable, TypeVariable]
+    
+    def renew(tv: TypeVariable): TypeVariable =
+      if (tv.lowerBounds.isEmpty && tv.upperBounds.isEmpty) tv
+      else renewed.get(tv) match {
+        case S(tv2) => tv2
+        case N =>
+          val tv2 = freshVar(tv.prov, tv.nameHint)(0)
+          renewed += tv -> tv2
+          tv2.lowerBounds = tv.lowerBounds.map(go(_, true))
+          tv2.upperBounds = tv.upperBounds.map(go(_, false))
+          tv2
+      }
+    
+    def go(st: SimpleType, pol: Bool): SimpleType = st match {
+      case ExtrType(_) => st
+      case tv: TypeVariable => renew(tv)
+      case NegType(und) => go(und, !pol).neg()
+      case TypeBounds(lb, ub) => if (pol) go(ub, true) else go(lb, false)
+      case RecordType(fs) => RecordType(fs.mapValues(go(_, pol)))(st.prov)
+      case TupleType(fs) => TupleType(fs.mapValues(go(_, pol)))(st.prov)
+      case FunctionType(l, r) => FunctionType(go(l, !pol), go(r, pol))(st.prov)
+      case ProvType(underlying) => ProvType(go(underlying, pol))(st.prov)
+      case ProxyType(underlying) => go(underlying, pol)
+      case wo @ Without(base, names) => Without(go(base, pol), names)(wo.prov)
+      case tr @ TypeRef(defn, targs) => tr.copy(targs = targs.map { targ =>
+          TypeBounds.mk(go(targ, false), go(targ, true), targ.prov)
+        })(tr.prov, tr.ctx)
+      case ty @ ComposedType(true, l, r) => go(l, pol) | go(r, pol)
+      case ty @ (ComposedType(false, _, _) | _: ObjectTag) =>
+        val dnf @ DNF(cs) = DNF.mk(ty, pol)
+        cs.sorted.map { c =>
+          c.copy(vars = c.vars.map(renew), nvars = c.nvars.map(renew)).toTypeWith(_ match {
+            case LhsRefined(bo, tts, rcd) =>
+              lazy val newRcd = rcd.copy(
+                rcd.fields.filterNot(_._1.name.isCapitalized).mapValues(go(_, pol)))(rcd.prov)
+              bo match {
+                case S(cls @ ClassTag(Var(tagNme), ps)) if !primitiveTypes.contains(tagNme) =>
```

Comment:
Should we also handle traits?

Reply:
In principle, it would be nice. But it would be a lot of work and may not be that important. The thing that makes the situation with traits less straightforward is that we don't have a single-inheritance hierarchy for traits, so there could be many traits mixed in, some related with other. For example if trait `T` inherits from class `C[int]` then we should probably reconstruct `c & t & { x: 1 }` as `T & { x: 1 }`.